sdv/sampling/hierarchical_sampler.py

```diff
@@ -123,7 +117,33 @@ def _add_child_rows(self, child_name, parent_name, parent_row, sampled_data, num
                 sampled_data[child_name] = pd.concat(
                     [previous, sampled_rows]).reset_index(drop=True)
 
-    def _sample_children(self, table_name, sampled_data):
+    def _enforce_table_sizes(self, child_name, table_name, scale, sampled_data):
+        total_num_rows = int(self._table_sizes[child_name] * scale)
+        for foreign_key in self.metadata._get_foreign_keys(table_name, child_name):
+            num_rows_key = f'__{child_name}__{foreign_key}__num_rows'
+            min_rows = self._min_child_rows[num_rows_key]
+            max_rows = self._max_child_rows[num_rows_key]
+            key_data = sampled_data[table_name][num_rows_key].fillna(0).round()
+            sampled_data[table_name][num_rows_key] = key_data.clip(min_rows, max_rows)
+
+            while sum(sampled_data[table_name][num_rows_key]) != total_num_rows:
+                num_rows_column = sampled_data[table_name][num_rows_key].argsort()
+                if sum(sampled_data[table_name][num_rows_key]) < total_num_rows:
+                    for i in num_rows_column:
+                        if sampled_data[table_name][num_rows_key][i] >= max_rows:
+                            break
+                        sampled_data[table_name][num_rows_key][i] += 1
+                        if sum(sampled_data[table_name][num_rows_key]) == total_num_rows:
+                            break
+                else:
+                    for i in num_rows_column[::-1]:
+                        if sampled_data[table_name][num_rows_key][i] <= min_rows:
+                            break
+                        sampled_data[table_name][num_rows_key][i] -= 1
```

Comment:
This line seems to be emitting a warning:
```
[/Users/frances/Documents/SDV-clean/sdv/sampling/hierarchical_sampler.py:142]: FutureWarning: ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy

  sampled_data[table_name][num_rows_key][i] -= 1
```